packages/opal-client/opal_client/policy_store/opa_client.py

```diff
@@ -741,10 +753,60 @@ async def set_policy_data(
         async with aiohttp.ClientSession() as session:
             try:
                 headers = await self._get_auth_headers()
+                data = json.dumps(
+                    policy_data,
+                    default=custom_encoder(by_alias=True, exclude_none=True),
+                )
 
                 async with session.put(
                     f"{self._opa_url}/data{path}",
-                    data=json.dumps(policy_data, default=str),
+                    data=data,
+                    headers=headers,
+                    **self._ssl_context_kwargs,
+                ) as opa_response:
+                    response = await proxy_response_unless_invalid(
+                        opa_response,
+                        accepted_status_codes=[
+                            status.HTTP_204_NO_CONTENT,
+                            status.HTTP_304_NOT_MODIFIED,
+                        ],
+                    )
+                    if self._policy_data_cache:
+                        self._policy_data_cache.set(path, json.loads(data))
+                    return response
+            except aiohttp.ClientError as e:
+                logger.warning("Opa connection error: {err}", err=repr(e))
+                raise
+
+    @affects_transaction
+    @retry(**RETRY_CONFIG)
+    async def patch_policy_data(
+        self,
+        policy_data: List[JSONPatchAction],
+        path: str = "",
+        transaction_id: Optional[str] = None,
+    ):
+        path = self._safe_data_module_path(path)
+
+        # in OPA, the root document must be an object, so we must wrap list values
+        if not path and isinstance(policy_data, list):
+            logger.warning(
+                "OPAL client was instructed to put a list on OPA's root document. In OPA the root document must be an object so the original value was wrapped."
+            )
+            policy_data = {"items": policy_data}
+
+        async with aiohttp.ClientSession() as session:
+            try:
+                headers = await self._get_auth_headers()
+                headers["Content-Type"] = "application/json-patch+json"
+                data = json.dumps(
```

Comment:
Seems like that `json.dumps` with the same parameters repeats itself.
Maybe wrap it in a utility function? (or do the encoding upper the call tree in a common ancestor if possible?)
Also - a comment on why those parameters are needed would be nice.

Reply:
Reasons for the parameters are
`by_alias` - the `from_field` attribute in the model `JSONPatchAction` is loaded using the field name and by default fastAPI would use the alias name while [returning the response ](https://github.com/tiangolo/fastapi/discussions/2753#discussioncomment-4077051) but as we are not returning the response but doing a json dump we need to explicitly set `by_alias` to `True` in pydantic
`exclude_none` - the `JSONPatchAction` has optional fields `value` and `from` which will be intialized to `None`, but we would not want these attributes to be present while sending the request to OPA or while doing a patch action on the offline data when offline mode is enabled (although in both cases having those fields won't result in an error), this is more important for the `set_policy_data`  where let's say one wants to store a list of `JSONPatchAction` data, in this case the default fields shouldn't be present in the data that will be written to OPA

Reply:
The reason for the `json.dumps` repeating in the `patch_policy_data` and the `OpaStaticDataCache.patch` method is that the `path` attribute in the JSON patch request is the relative path, the `DataSourceEntry.dst_path`+`path` would make the actual path in the JSON document of OPA 
so to do a [patch operation](https://python-json-patch.readthedocs.io/en/latest/tutorial.html#applying-a-patch) on the offline data (JSON file) we would need the actual path, basically the path attribute in every JSON patch action would be the relative path for OPA (original request payload) and it would be the actual path for editing the offline mode data file